modules/extension/wmts/src/main/java/org/geotools/ows/wmts/map/WMTSCoverageReader.java

```diff
@@ -563,4 +564,141 @@ public String getRequestedTime() {
+            throw new RuntimeException(factoryException);
+        }
+    }
+
+    private String errorMessage(CoordinateReferenceSystem targetCrs) {
+        StringBuilder msg =
+                new StringBuilder("Something wrong happended while trying to reproject ");
+        if (coverageName != null) msg.append(coverageName).append(" ");
+        msg.append("WMTS coverage to ");
+        if (crs != null) msg.append(CRS.toSRS(targetCrs));
+        else msg.append(" null srs.");
+        return msg.toString();
+    }
+
+    private GridCoverage2D getMapReproject(
+            ReferencedEnvelope requestedEnvelope, String time, WMTSReadParameters readParameters)
+            throws IOException {
+        try {
+            CoordinateReferenceSystem targetCRS = requestedEnvelope.getCoordinateReferenceSystem();
+            ReferencedEnvelope nativeEnvelope =
+                    requestedEnvelope.transform(getBestSourceCRS(targetCRS, readParameters), false);
+            ReferencedEnvelope gridEnvelope =
+                    initTileRequest(
+                            nativeEnvelope,
+                            readParameters.getWidth(),
+                            readParameters.getHeight(),
+                            time);
+            GridCoverage2D result = getMap(nativeEnvelope, gridEnvelope);
+            if (!CRS.equalsIgnoreMetadata(result.getCoordinateReferenceSystem(), targetCRS))
+                result =
+                        reprojectManually(
+                                result, new GeneralEnvelope(requestedEnvelope), readParameters);
+            return result;
+        } catch (FactoryException | TransformException e) {
+            if (LOGGER.isLoggable(Level.SEVERE))
+                LOGGER.log(
+                        Level.SEVERE,
+                        "Error while reprojecting the requested envelope to the selected native crs.",
+                        e);
+            throw new IOException(e);
+        }
+    }
+
+    // get the CRS to be used to reproject to a CRS that the server doesn't support.
+    CoordinateReferenceSystem getBestSourceCRS(
+            CoordinateReferenceSystem targetCRS, WMTSReadParameters readParameters)
+            throws IOException {
+        try {
+            CoordinateReferenceSystem result = null;
+            String switched = switchDefinition(targetCRS);
+            if (switched != null) {
+                // check first by changing the srs definition
+                result = CRS.decode(switched);
+            } else if (readParameters.getSourceCRS() != null) {
+                // otherwise get the preferred one
+                result = readParameters.getSourceCRS();
+            } else if (!validSRS.isEmpty()) {
+                // last attempt is to retrieve the firs valid srs available
+                LOGGER.fine(() -> "Retrieving random srs to perform reprojection");
+                String srs = validSRS.iterator().next();
```

Comment:
This could likely be more robust. What if the remote server first SRS cannot be decoded?
I'd loop with a try/catch inside, and try them all, before throwing the towel.

Reply:
fixed